### ISSUE TITLE
Removal of Unmodifiable classes broke dart2wasm tests

### ISSUE THREAD

**Issue by aam** [dart-lang/sdk#56014]
https://github.com/dart-lang/sdk/commit/18994e6e46ec9fb2fac8368c43d448119abd579f [broke ](https://logs.chromium.org/logs/dart/buildbucket/cr-buildbucket/8745207616937421793/+/u/test_results/new_test_failures__logs_)
* lib/typed_data/polymorphic_unmodifiable_typed_data_test
* lib/typed_data/unmodifiable_typed_data_simd_test
* lib/typed_data/unmodifiable_typed_data_test

on dart2wasm-linux-jscm-chrome Pass -> RuntimeError (expected Pass)

```

/=====================================================================================================\
| lib/typed_data/polymorphic_unmodifiable_typed_data_test broke (Pass -> RuntimeError, expected Pass) |
\=====================================================================================================/

--- Command "dart2wasm" (took 792ms):
DART_CONFIGURATION=ReleaseX64 CHROME_PATH=/b/s/w/ir/third_party/browsers/chrome/chrome/google-chrome pkg/dart2wasm/tool/compile_benchmark -O0 --js-compatibility --compiler-asserts -Dtest_runner.configuration=dart2wasm-linux-jscm-chrome /b/s/w/ir/tests/lib/typed_data/polymorphic_unmodifiable_typed_data_test.dart /b/s/w/ir/out/ReleaseX64/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_polymorphic_unmodifiable_typed_data_test/polymorphic_unmodifiable_typed_data_test.wasm

exit code:
0

--- Command "chrome" (took 327ms):
/b/s/w/ir/tools/sdks/dart-sdk/bin/dart pkg/test_runner/bin/launch_browser.dart chrome http://127.0.0.1:38493/root_build/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_polymorphic_unmodifiable_typed_data_test/test.html?crossOriginPort=43311

Runtime error:
[object WebAssembly.Exception]
<undefined stack>

exit code:
0

--- To debug locally, run:
/b/s/w/ir/tools/sdks/dart-sdk/bin/dart /b/s/w/ir/pkg/test_runner/bin/http_server.dart -p 38493 -c 43311 --network 127.0.0.1 --build-directory=/b/s/w/ir/out/ReleaseX64/ --runtime=chrome --packages=/b/s/w/ir/.dart_tool/package_config.json

--- Re-run this test:
python3 tools/test.py -n dart2wasm-linux-jscm-chrome lib/typed_data/polymorphic_unmodifiable_typed_data_test


/==============================================================================================\
| lib/typed_data/unmodifiable_typed_data_simd_test broke (Pass -> RuntimeError, expected Pass) |
\==============================================================================================/

--- Command "dart2wasm" (took 807ms):
DART_CONFIGURATION=ReleaseX64 CHROME_PATH=/b/s/w/ir/cache/builder/sdk/third_party/browsers/chrome/chrome/google-chrome pkg/dart2wasm/tool/compile_benchmark -O0 --js-compatibility --compiler-asserts -Dtest_runner.configuration=dart2wasm-linux-jscm-chrome /b/s/w/ir/cache/builder/sdk/tests/lib/typed_data/unmodifiable_typed_data_simd_test.dart /b/s/w/ir/cache/builder/sdk/out/ReleaseX64/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_unmodifiable_typed_data_simd_test/unmodifiable_typed_data_simd_test.wasm

exit code:
0

--- Command "chrome" (took 58ms):
/b/s/w/ir/cache/builder/sdk/tools/sdks/dart-sdk/bin/dart pkg/test_runner/bin/launch_browser.dart chrome http://127.0.0.1:38895/root_build/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_unmodifiable_typed_data_simd_test/test.html?crossOriginPort=33223

Runtime error:
RuntimeError: illegal cast
RuntimeError: illegal cast
    at new NaiveUnmodifiableInt32x4List (initializer) (http://127.0.0.1:38895/root_build/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_unmodifiable_typed_data_simd_test/unmodifiable_typed_data_simd_test.wasm:wasm-function[1020]:0x143d7)
    at NaiveUnmodifiableInt32x4List (http://127.0.0.1:38895/root_build/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_unmodifiable_typed_data_simd_test/unmodifiable_typed_data_simd_test.wasm:wasm-function[1019]:0x143b4)
    at JSInt32x4ArrayImpl.asUnmodifiableView (http://127.0.0.1:38895/root_build/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_unmodifiable_typed_data_simd_test/unmodifiable_typed_data_simd_test.wasm:wasm-function[993]:0x13cd6)
    at int32x4Test (http://127.0.0.1:38895/root_build/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_unmodifiable_typed_data_simd_test/unmodifiable_typed_data_simd_test.wasm:wasm-function[716]:0xff93)
    at main (http://127.0.0.1:38895/root_build/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_unmodifiable_typed_data_simd_test/unmodifiable_typed_data_simd_test.wasm:wasm-function[87]:0x85f7)
    at main tear-off trampoline (http://127.0.0.1:38895/root_build/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_unmodifiable_typed_data_simd_test/unmodifiable_typed_data_simd_test.wasm:wasm-function[89]:0x8610)
    at _invokeMain (http://127.0.0.1:38895/root_build/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_unmodifiable_typed_data_simd_test/unmodifiable_typed_data_simd_test.wasm:wasm-function[92]:0x8826)
    at Module.invoke (http://127.0.0.1:38895/root_build/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_unmodifiable_typed_data_simd_test/unmodifiable_typed_data_simd_test.mjs:548:28)
    at http://127.0.0.1:38895/root_build/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_unmodifiable_typed_data_simd_test/test.html?crossOriginPort=33223:29:23
    at dartMainRunner (http://127.0.0.1:38895/root_dart/pkg/test_runner/lib/src/test_controller.js:319:5)

Deobfuscated error and stack:
RuntimeError: illegal cast
RuntimeError: illegal cast
    at     at new NaiveUnmodifiableInt32x4List (initializer) (out/ReleaseX64/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_unmodifiable_typed_data_simd_test/unmodifiable_typed_data_simd_test.wasm:wasm-function[1020]:0x143d7) unparsed
    at     at NaiveUnmodifiableInt32x4List (out/ReleaseX64/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_unmodifiable_typed_data_simd_test/unmodifiable_typed_data_simd_test.wasm:wasm-function[1019]:0x143b4)                   unparsed
    at     at JSInt32x4ArrayImpl.asUnmodifiableView (out/ReleaseX64/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_unmodifiable_typed_data_simd_test/unmodifiable_typed_data_simd_test.wasm:wasm-function[993]:0x13cd6)           unparsed
    at     at int32x4Test (out/ReleaseX64/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_unmodifiable_typed_data_simd_test/unmodifiable_typed_data_simd_test.wasm:wasm-function[716]:0xff93)                                      unparsed
    at     at main (out/ReleaseX64/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_unmodifiable_typed_data_simd_test/unmodifiable_typed_data_simd_test.wasm:wasm-function[87]:0x85f7)                                              unparsed
    at     at main tear-off trampoline (out/ReleaseX64/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_unmodifiable_typed_data_simd_test/unmodifiable_typed_data_simd_test.wasm:wasm-function[89]:0x8610)                          unparsed
    at     at _invokeMain (out/ReleaseX64/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_unmodifiable_typed_data_simd_test/unmodifiable_typed_data_simd_test.wasm:wasm-function[92]:0x8826)                                       unparsed
    at Module.invoke                                                                                                                                                                                                                                out/ReleaseX64/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_unmodifiable_typed_data_simd_test/unmodifiable_typed_data_simd_test.mjs 548:28
    at <fn>                                                                                                                                                                                                                                         out/ReleaseX64/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_unmodifiable_typed_data_simd_test/test.html 29:23
    at dartMainRunner                                                                                                                                                                                                                               pkg/test_runner/lib/src/test_controller.js 319:5


exit code:
0

--- To debug locally, run:
/b/s/w/ir/cache/builder/sdk/tools/sdks/dart-sdk/bin/dart /b/s/w/ir/cache/builder/sdk/pkg/test_runner/bin/http_server.dart -p 38895 -c 33223 --network 127.0.0.1 --build-directory=/b/s/w/ir/cache/builder/sdk/out/ReleaseX64/ --runtime=chrome --packages=/b/s/w/ir/cache/builder/sdk/.dart_tool/package_config.json

--- Re-run this test:
python3 tools/test.py -n dart2wasm-linux-jscm-chrome lib/typed_data/unmodifiable_typed_data_simd_test


/=========================================================================================\
| lib/typed_data/unmodifiable_typed_data_test broke (Pass -> RuntimeError, expected Pass) |
\=========================================================================================/

--- Command "dart2wasm" (took 720ms):
DART_CONFIGURATION=ReleaseX64 CHROME_PATH=/b/s/w/ir/third_party/browsers/chrome/chrome/google-chrome pkg/dart2wasm/tool/compile_benchmark -O0 --js-compatibility --compiler-asserts -Dtest_runner.configuration=dart2wasm-linux-jscm-chrome /b/s/w/ir/tests/lib/typed_data/unmodifiable_typed_data_test.dart /b/s/w/ir/out/ReleaseX64/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_unmodifiable_typed_data_test/unmodifiable_typed_data_test.wasm

exit code:
0

--- Command "chrome" (took 76ms):
/b/s/w/ir/tools/sdks/dart-sdk/bin/dart pkg/test_runner/bin/launch_browser.dart chrome http://127.0.0.1:38493/root_build/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_unmodifiable_typed_data_test/test.html?crossOriginPort=43311

Runtime error:
[object WebAssembly.Exception]
<undefined stack>

exit code:
0

--- To debug locally, run:
/b/s/w/ir/tools/sdks/dart-sdk/bin/dart /b/s/w/ir/pkg/test_runner/bin/http_server.dart -p 38493 -c 43311 --network 127.0.0.1 --build-directory=/b/s/w/ir/out/ReleaseX64/ --runtime=chrome --packages=/b/s/w/ir/.dart_tool/package_config.json

--- Re-run this test:
python3 tools/test.py -n dart2wasm-linux-jscm-chrome lib/typed_data/unmodifiable_typed_data_test

```

**Comment by osa1** [comment]
@mkustermann I think it may make sense to run jscm test config before submitting patches. That configuration greatly increases test coverage. This issue was caused because we didn't run it when submitting.

**Comment by rakudrama** [comment]
+1 on adding jscm. I had no idea that not all the relevant tests were run.

This fixes the one with the stack trace: https://dart-review.googlesource.com/c/sdk/+/371743

I'm not sure where to start with the other two that just report:
```
Runtime error:
[object WebAssembly.Exception]
<undefined stack>
```

**Comment by mkustermann** [comment]
> @mkustermann I think it may make sense to run jscm test config before submitting patches. That configuration greatly increases test coverage. This issue was caused because we didn't run it when submitting.

Ideally all CI bots also run on CQ. If we don't have resources and we had to pick, I'd pick not running JSCM because we don't have any customers of that mode atm. /cc @athomas does it increase CQ times if we add another builder?

> I'm not sure where to start with the other two that just report:

You can often get the real stacks by running with `D8` instead of `Chrome`. So in this case, if I run
```
% python3 tools/test.py -n dart2wasm-linux-jscm-d8 lib/typed_data/polymorphic_unmodifiable_typed_data_test
...
stdout:
Attempt to execute code removed by Dart AOT compiler (TFA)
    at Error._throwWithCurrentStackTrace (wasm://wasm/00086132:wasm-function[212]:0xa9dc)
    at new NaiveUnmodifiableInt32x4List (initializer) (wasm://wasm/00086132:wasm-function[925]:0x13c74)
    at NaiveUnmodifiableInt32x4List (wasm://wasm/00086132:wasm-function[924]:0x13c4d)
    at JSInt32x4ArrayImpl.asUnmodifiableView (wasm://wasm/00086132:wasm-function[914]:0x13aa9)
    at testInt32x4 (wasm://wasm/00086132:wasm-function[731]:0x11204)
    at main (wasm://wasm/00086132:wasm-function[94]:0x8eb8)
    at main tear-off trampoline (wasm://wasm/00086132:wasm-function[96]:0x8ed5)
    at _invokeMain (wasm://wasm/00086132:wasm-function[99]:0x90eb)
/usr/local/google/home/kustermann/repositories/sdk-gclient4/sdk/pkg/dart2wasm/bin/run_wasm.js:343: [object WebAssembly.Exception]
  }
  ^
```

Looking at the code, it may be due to this:
```dart
final class NaiveUnmodifiableInt32x4List extends NaiveInt32x4List {
  NaiveUnmodifiableInt32x4List(Int32x4List list)
      : super.externalStorage(unsafeCast<NaiveInt32x4List>(list)._storage);
  ...
}
```
TFA may infer that `list` is always a `JSInt32x4ArrayImpl` and so it knows that this `unsafeCast<NaiveInt32x4List>(list)`  cannot be right, so it inserts a throw.

**Comment by athomas** [comment]
Adding dart2wasm-linux-jscm-chrome to the CQ SGTM, it runs relatively quick and should mostly trigger on dart2wasm related changes.

**Comment by osa1** [comment]
Fixed by the commit referenced above. 08ca6316738f6d9a89dc35b226599d702257fc29 removes unsafe casts in the file so that we won't have this issue again.

JSCM added to the CQ tests in https://dart-review.googlesource.com/c/sdk/+/371542.

**Comment by mkustermann** [comment]
The test is still failing in JSCM mode (but now due to a different reason):
```
% tools/build.py -nvh -d0 -mrelease -ax64 dart2wasm  create_sdk 
...
% python3 tools/test.py -n dart2wasm-linux-jscm-chrome lib/typed_data/polymorphic_unmodifiable_typed_data_test
                                                                                                                                                                                                                                                           
FAILED: dart2wasm-chrome release_x64 lib/typed_data/polymorphic_unmodifiable_typed_data_test                                                                                                                                                               
Expected: Pass                                                                                                                                                                                                                                             
Actual: RuntimeError                                                                                                                                                                                                                                       
...                                                      
Runtime error:                                             
[object WebAssembly.Exception]                                                                                                                                                                                                                             
<undefined stack>                                                                                                                                                                                                                                          
                                                                                                                             
Events:                                                                                                                                                                                                                                                    
- debug "test_controller.js started"                                                                                         
- debug "Sending events to driver page (isFirstMessage = true, isStatusUpdate = true, isDone = false)"                                                                                                                                                     
- print "dart-calling-main"                            
- print "UnimplementedError"                                                                                                                                                                                                                               
- print "    at Error._throwWithCurrentStackTrace (http://127.0.0.1:36223/root_build/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_polymorphic_unmodifiable_typed_data_test/polymorphic_unmodifiable_typed_data_test.wasm:wasm-function[
212]:0xaa5f)                                                                                                                                                                                                                                               
    at JSDataViewImpl.asUnmodifiableView (http://127.0.0.1:36223/root_build/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_polymorphic_unmodifiable_typed_data_test/polymorphic_unmodifiable_typed_data_test.wasm:wasm-function[740]:0x11
6fc)                                                                                                                         
    at testByteData (http://127.0.0.1:36223/root_build/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_polymorphic_unmodifiable_typed_data_test/polymorphic_unmodifiable_typed_data_test.wasm:wasm-function[734]:0x115de)
    at main (http://127.0.0.1:36223/root_build/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_polymorphic_unmodifiable_typed_data_test/polymorphic_unmodifiable_typed_data_test.wasm:wasm-function[94]:0x8f47)
    at main tear-off trampoline (http://127.0.0.1:36223/root_build/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_polymorphic_unmodifiable_typed_data_test/polymorphic_unmodifiable_typed_data_test.wasm:wasm-function[96]:0x8f58)
    at _invokeMain (http://127.0.0.1:36223/root_build/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_polymorphic_unmodifiable_typed_data_test/polymorphic_unmodifiable_typed_data_test.wasm:wasm-function[99]:0x916e)
    at Module.invoke (http://127.0.0.1:36223/root_build/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_polymorphic_unmodifiable_typed_data_test/polymorphic_unmodifiable_typed_data_test.mjs:558:28)
    at http://127.0.0.1:36223/root_build/generated_tests/dart2wasm-linux-jscm-chrome/tests_lib_typed_data_polymorphic_unmodifiable_typed_data_test/test.html?crossOriginPort=37871:29:23"
- {                                                                                                                          
      "type": "sync_exception",                                                                                                                                                                                                                            
      "value": "[object WebAssembly.Exception]",                                                                                                                                                                                                           
      "timestamp": 2.15,                       
      "stack_trace": "<undefined stack>"
}                             
```

which seems to be due to
```
 final class JSDataViewImpl implements ByteData {
  ...
  @override
  ByteData asUnmodifiableView() => throw UnimplementedError();
}
```

**Comment by osa1** [comment]
I'm implemented unmodifiable JS data view view in https://dart-review.googlesource.com/c/sdk/+/372082.

**Comment by osa1** [comment]
Missing `asUnmodifiableView` was implemented in 3b5065b49e118733a2e7671820e6e4d2c8e0d51d.
